sys/xtimer/xtimer.c

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ * @ingroup xtimer
+ * @{
+ * @file
+ * @brief xtimer convenience functionality
+ * @author Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+#include <stdint.h>
+#include <string.h>
+
+#include "xtimer.h"
+#include "mutex.h"
+#include "thread.h"
+#include "irq.h"
+
+#include "timex.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static void _callback_unlock_mutex(void* arg)
+{
+    mutex_t *mutex = (mutex_t *) arg;
+    mutex_unlock(mutex);
+}
+
+void _xtimer_sleep(uint32_t offset, uint32_t long_offset)
+{
+    xtimer_t timer;
+    mutex_t mutex = MUTEX_INIT;
+
+    timer.callback = _callback_unlock_mutex;
+    timer.arg = (void*) &mutex;
+
+    mutex_lock(&mutex);
+    _xtimer_set64(&timer, offset, long_offset);
+    mutex_lock(&mutex);
+}
+
+int xtimer_usleep_until(uint32_t *last_wakeup, uint32_t interval) {
+    int ret = 0;
+    xtimer_t timer;
+    mutex_t mutex = MUTEX_INIT;
+
+    timer.callback = _callback_unlock_mutex;
+    timer.arg = (void*) &mutex;
+
+    uint32_t target = *last_wakeup + interval;
+
+    uint32_t now = xtimer_now();
+    /* make sure we're not setting a value in the past */
+    if (now < *last_wakeup) {
+        /* base timer overflowed */
+        if (!((target < *last_wakeup) && (target > now))) {
+            goto out;
```

Comment:
I also like to avoid as many indentations as possible (e.g. see https://github.com/RIOT-OS/RIOT/pull/2926#discussion_r37213384), but I also tend to overlook jump marks in C code, because they are so rare and syntactically so different from the rest.